rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,742 @@
+  // The matched Foo resource.
+  Foo foo = 1;
+}
+```
+
+#### List
+
+The `List` RPC takes the requested page size and starting point and returns a list of resources that match. If there are
+additional resources, the response MUST also include a token that indicates where the next page of results begins.
+
+Most legacy APIs do not provide a paginated way to retrieve resources and instead offer some kind of `GetAllFoos` RPC
+which either returns all `Foo` in a single message or leverages a server side stream to send each `Foo` one at a time.
+Returning all items in a single message causes problems when the number of resources scales beyond gRPC message size
+limits. To provide parity with this legacy API if needed, a helper method should be implemented on the client which
+builds the entire resource set by repeatedly calling `List` until all pages have been consumed.
+
+```protobuf
+// Returns a page of Foo and the token to find the next page of items.
+    rpc ListFoos(ListFoosRequest) returns (ListFoosResponse);
+
+message ListFoosRequest {
+  // The maximum number of items to return.
+  // The server may impose a different page size at its discretion.
+  int32 page_size = 1;
+  // The next_page_token value returned from a previous List request, if any.
+  string page_token = 2;
+}
+
+message ListFoosResponse {
+  // The page of Foo that matched the request.
+  repeated Foo foos = 1;
+  // Token to retrieve the next page of results, or empty if there are no
+  // more results in the list.
+  string next_page_token = 2;
+}
+```
+
+A listing operation should not abort entirely if a single item cannot be (un)marshalled, it should instead be logged,
+and the rest of the page should be processed. Aborting an entire page when a single entry is invalid causes the cache
+to be permanently unhealthy since it is never able to initialize loading the affected resource.
+
+#### Delete
+
+The `Delete` RPC takes the parameters required to match a resource and performs a hard delete of the specified resource
+from the backend and returns a `google.protobuf.Empty`.
+
+The request MUST fail and return a `trace.NotFound` error if there is no matching resource in the backend.
+
+```protobuf
+// Remove a matching Foo resource
+    rpc DeleteFoo(DeleteFooRequest) returns (google.protobuf.Empty);
+
+message DeleteFooRequest {
+  // Name of the foo to remove. Some resource may require more parameters to match and
+  // may not use the name at all.
+  string foo_id = 1;
+}
+```
+
+### Backend Storage
+
+A backend service to handle persisting and retrieving a resource from the backend is typically defined in
+`lib/services/local/foo.go`. An accompanying interface which mirrors the service is defined in `lib/services/foo.go`.
+Continuing on with the example above, the backend service for the `Foo` resource might look like the following.
+
+The sections below contain a reference example for how to interact with the backend to perform common operations on a
```

Comment:
```suggestion
Continuing on with the example above, the sections below show how the backend service for the `Foo` resource might look like.

The sections also contain a reference example for how to interact with the backend to perform common operations on a
```

At the moment the text says "following" but nothing follows the paragraph, making it seem like something was omitted by mistake.